lib/improver/spotdata_new/apply_lapse_rate.py

```diff
@@ -0,0 +1,206 @@
+        except TypeError:
+            match = ((attributes.viewitems() &
+                      reference_attributes.viewitems()) ==
+                     reference_attributes.viewitems())
+        if match is not True:
+            raise ValueError('Cubes do not share the metadata identified '
+                             'by the grid_metadata_identifier ({})'.format(
+                                 self.grid_metadata_identifier))
+
+    def check_grid_match(self, cubes):
+        """
+        Uses the provided grid_metadata_identifier to extract and compare
+        attributes on the input cubes. The expectation is that all the metadata
+        identified should match for the cubes to be deemed compatible.
+
+        Args:
+            cubes (list of iris.cube.Cube items):
+                List of cubes for which the attributes should be tested.
+        """
+        # Allow user to bypass cube comparison by setting identifier to None.
+        if self.grid_metadata_identifier is None:
+            return
+
+        reference_attributes = self._get_attributes(cubes[0])
+        for cube in cubes:
+            attributes = self._get_attributes(cube)
+            self._compare_attributes(attributes, reference_attributes)
+
+    def process(self, spot_data_cube, neighbour_cube, gridded_lapse_rate_cube):
+        """
+        Extract lapse rates from the appropriate grid points and apply them to
+        the spot extracted temperatures.
+
+        The calculation is::
+
+         lapse_rate_adjusted_temperatures = temperatures + lapse_rate *
+         vertical_displacement
+
+        Args:
+            spot_data_cube (iris.cube.Cube):
+                The spot data cube of temperatures extracted from the gridded
+                fields. These temperatures will have been extracted using the
+                same neighbour_cube and neighbour_selection_method that are
+                being used here.
+            neighbour_cube (iris.cube.Cube):
+                The neighbour_cube that contains the grid coordinates at which
+                lapse rates should be extracted and the vertical displacement
+                between those grid points on the model orography and the spot
+                data sites actual altitudes.
+            gridded_lapse_rate_cube (iris.cube.Cube):
+                A cube of temperature lapse rates on the same grid as that from
+                which the spot data temperatures were extracted.
+        Returns:
+            new_spot_cube (iris.cube.Cube):
+                A copy of the input spot_data_cube with the data modified by
+                the lapse rates to give a better representation of the site's
+                temperatures.
+        """
+        # Check the cubes are compatible.
+        self.check_grid_match([spot_data_cube, neighbour_cube,
+                               gridded_lapse_rate_cube])
+
+        # Extract the lapse rates that correspond to the spot sites.
```

Comment:
`Extract lapse rate for each site (in spot_data_cube)`?

Reply:
I do not prefer this. Given that above we have tried to emphasise that the temperature data is for multiple sites, removing all the plurals from the sentence seems counter intuitive.

Reply:
Yes, there are data for several different spot sites in the same cube.  I think this confused me when first working through this PR - it's not immediately obvious to someone who doesn't know the design of our spot extraction code that we would process all spot sites at once, rather than as separate cubes.  However if the docstring is clear on this I am happy to leave the inline comments as they are.